databuilder/query_model/transforms.py

```diff
@@ -27,9 +28,64 @@
     get_input_nodes,
     get_series_type,
     get_sorts,
-    validate_node,
 )
-from databuilder.utils.collections_utils import DefaultIdentityDict, IdentitySet
+
+
+class QueryGraphRewriter:
+    """
+    Query graphs are composed of immutable query model Node objects. This class allows
+    us to modify these graphs by building up a set of replacements (node X replaces node
+    A; node Y replaces node B etc) and then creating a new graph based on an existing
+    one but incorporating these replacements.
+    """
+
+    def __init__(self):
+        self.replacements = {}
+
+    def replace(self, target_node, new_node):
+        self.replacements[target_node] = new_node
+
+    def rewrite(self, obj, replacing=frozenset()):
+        if isinstance(obj, Node):
+            return self.rewrite_node(obj, replacing)
+        elif isinstance(obj, dict):
+            # Dicts need rewriting because they may contain references to other nodes
+            return {
+                self.rewrite(k, replacing): self.rewrite(v, replacing)
+                for k, v in obj.items()
+            }
+        elif isinstance(obj, frozenset):
+            # As do frozensets
+            return frozenset(self.rewrite(v, replacing) for v in obj)
+        else:
+            # Any other values in the query graph we return unchanged
+            return obj
```

Comment:
Is it worth being explicit about the kinds of things that `obj` can be in this branch, and then adding an `else: assert False` to catch new kinds of object that we might want to handle differently?

Reply:
Yes, I originally didn't do this because I didn't want this code to know about every sort of thing we might wrap in a `Value` node. But now you raise this I think a better approach would be to handle `Value` nodes explicitly and then exhaustively list the other types of thing expect.

Reply:
I've refactored it along these lines, and ended up pulling the rewriter out into its own module. This is all in additional commits; I haven't tried to rebase these.